pkg/visor/init.go

```diff
@@ -0,0 +1,432 @@
+		skyenv.VPNServerName: func() []string { return makeVPNEnvs(v.conf, v.net) },
+	})
+
+	v.procM = procM
+	v.appL = launch
+	return report(nil)
+}
+
+func makeVPNEnvs(conf *Config, n *snet.Network) []string {
+	var envCfg vpn.DirectRoutesEnvConfig
+
+	if conf.Dmsg != nil {
+		envCfg.DmsgDiscovery = conf.Dmsg.Discovery
+		envCfg.DmsgServers = n.Dmsg().ConnectedServers()
+	}
+	if conf.Transport != nil {
+		envCfg.TPDiscovery = conf.Transport.Discovery
+	}
+	if conf.Routing != nil {
+		envCfg.RF = conf.Routing.RouteFinder
+	}
+	if conf.UptimeTracker != nil {
+		envCfg.UptimeTracker = conf.UptimeTracker.Addr
+	}
+	if conf.STCP != nil && len(conf.STCP.PKTable) != 0 {
+		envCfg.STCPTable = conf.STCP.PKTable
+	}
+
+	envMap := vpn.AppEnvArgs(envCfg)
+
+	envs := make([]string, 0, len(envMap))
+	for k, v := range vpn.AppEnvArgs(envCfg) {
+		envs = append(envs, fmt.Sprintf("%s=%s", k, v))
+	}
+	return envs
+}
+
+func initCLI(v *Visor) bool {
+	report := v.makeReporter("cli")
+
+	if v.conf.CLIAddr == "" {
+		v.log.Info("'cli_addr' is not configured, skipping.")
+		return report(nil)
+	}
+
+	cliL, err := net.Listen("tcp", v.conf.CLIAddr)
+	if err != nil {
+		return report(err)
+	}
+
+	v.pushCloseStack("cli.listener", func() bool {
+		return report(cliL.Close())
+	})
+
+	rpcS, err := newRPCServer(v, "CLI")
+	if err != nil {
+		return report(fmt.Errorf("failed to start rpc server for cli: %w", err))
+	}
+	go rpcS.Accept(cliL) // We not not use sync.WaitGroup here as it will never return anyway.
+
+	return report(nil)
+}
+
+func initHypervisors(v *Visor) bool {
+	report := v.makeReporter("hypervisors")
+
+	hvErrs := make(map[cipher.PubKey]chan error, len(v.conf.Hypervisors))
+	for _, hv := range v.conf.Hypervisors {
+		hvErrs[hv.PubKey] = make(chan error, 1)
+	}
+
+	for hvPK, hvErrs := range hvErrs {
+		log := v.MasterLogger().PackageLogger("hypervisor_client").WithField("hypervisor_pk", hvPK)
+
+		addr := dmsg.Addr{PK: hvPK, Port: skyenv.DmsgHypervisorPort}
+		rpcS, err := newRPCServer(v, addr.PK.String()[:shortHashLen])
+		if err != nil {
+			return report(fmt.Errorf("failed to start RPC server for hypervisor %s: %w", hvPK, err))
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		wg := new(sync.WaitGroup)
+		wg.Add(1)
+
+		go func(hvErrs chan error) {
+			defer wg.Done()
+			ServeRPCClient(ctx, log, v.net, rpcS, addr, hvErrs)
+		}(hvErrs)
+
+		v.pushCloseStack("hypervisor."+hvPK.String()[:shortHashLen], func() bool {
+			cancel()
+			wg.Wait()
+			return true
+		})
+	}
+
+	return report(nil)
+}
+
+func initUptimeTracker(v *Visor) bool {
+	report := v.makeReporter("uptime_tracker")
+	conf := v.conf.UptimeTracker
+
+	if conf == nil {
+		v.log.Info("'uptime_tracker' is not configured, skipping.")
+		return true
+	}
+
+	ut, err := utclient.NewHTTP(conf.Addr, v.conf.KeyPair.PubKey, v.conf.KeyPair.SecKey)
+	if err != nil {
+		// TODO(evanlinjin): We should design utclient to retry automatically instead of returning error.
+		//return report(err)
```

Comment:
Can this line be removed?

Reply:
@nkryuchkov I've left it there to give an example of how it should function once the above `TODO` comment is implemented. 😅